server/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -326,70 +310,51 @@ private static List<ShardRouting> rankShardsAndUpdateStats(
         }
 
         // Retrieve which nodes we can potentially send the query to
-        final Set<String> nodeIds = getAllNodeIds(shards);
-        final Map<String, Optional<ResponseCollectorService.ComputedNodeStats>> nodeStats = getNodeStats(nodeIds, collector);
+        final Map<String, ResponseCollectorService.ComputedNodeStats> nodeStats = getNodeStats(shards, collector);
 
         // Retrieve all the nodes the shards exist on
         final Map<String, Double> nodeRanks = rankNodes(nodeStats, nodeSearchCounts);
-
         // sort all shards based on the shard rank
         ArrayList<ShardRouting> sortedShards = new ArrayList<>(shards);
-        Collections.sort(sortedShards, new NodeRankComparator(nodeRanks));
-
-        // adjust the non-winner nodes' stats so they will get a chance to receive queries
-        if (sortedShards.size() > 1) {
-            ShardRouting minShard = sortedShards.get(0);
-            // If the winning shard is not started we are ranking initializing
-            // shards, don't bother to do adjustments
-            if (minShard.started()) {
-                String minNodeId = minShard.currentNodeId();
-                Optional<ResponseCollectorService.ComputedNodeStats> maybeMinStats = nodeStats.get(minNodeId);
-                if (maybeMinStats.isPresent()) {
-                    adjustStats(collector, nodeStats, minNodeId, maybeMinStats.get());
-                    // Increase the number of searches for the "winning" node by one.
-                    // Note that this doesn't actually affect the "real" counts, instead
-                    // it only affects the captured node search counts, which is
-                    // captured once for each query in TransportSearchAction
-                    nodeSearchCounts.compute(minNodeId, (id, conns) -> conns == null ? 1 : conns + 1);
-                }
-            }
-        }
-
-        return sortedShards;
-    }
-
-    private static class NodeRankComparator implements Comparator<ShardRouting> {
-        private final Map<String, Double> nodeRanks;
-
-        NodeRankComparator(Map<String, Double> nodeRanks) {
-            this.nodeRanks = nodeRanks;
-        }
-
-        @Override
-        public int compare(ShardRouting s1, ShardRouting s2) {
-            if (s1.currentNodeId().equals(s2.currentNodeId())) {
+        sortedShards.sort((left, right) -> {
+            final String leftId = left.currentNodeId();
+            final String rightId = right.currentNodeId();
+            if (leftId.equals(rightId)) {
                 // these shards on the same node
                 return 0;
             }
-            Double shard1rank = nodeRanks.get(s1.currentNodeId());
-            Double shard2rank = nodeRanks.get(s2.currentNodeId());
+            Double shard1rank = nodeRanks.get(leftId);
+            Double shard2rank = nodeRanks.get(rightId);
             if (shard1rank != null) {
```

Comment:
Can use `Comparator.nullsFirst(Comparator.naturalOrder())` in some capacity here

Reply:
Turns out that one is quite slow because if it doesn't inline fully, it adds serious method call overhead + more allocations here which is probably worse in this specific case.

Reply:
Even if you put the `Comparator` instance somewhere static?

Reply:
Yea, same thing, the current JVM implementations aren't able to understand static-constant objects like that and compile them like a chain of method references. Maybe that'll happen at some point, but it seems like a hard problem for the JVM to solve.

Reply:
OK then, worth a comment saying so

Reply:
👍 added